File: adapters/zapo/src/send.ts
/**
 * Sending a stanza through `zapo`.
 *
 * The outbound half of the boundary. Frames cross in the same shape they do
 * inbound — the bytes a decoder consumes — so a captured envelope can be sent
 * back as it stands, and record and replay stop being separate features.
 *
 * `zapo` sends decoded nodes rather than bytes, so this decodes the frame on the
 * way out. That is the adapter's job precisely so a consumer never has to know
 * which shape its engine wants: the Rust adapter puts the format byte back and
 * hands over bytes, this one decodes, and the consumer writes the same code
 * against either.
 *
 * # What this is not
 *
 * Not request/response. `zapo` can correlate a reply — `queryWithContext` — and
 * that is a separate capability (`l0.request`) an adapter claims on its own.
 * Writing to the socket and being handed the answer are different powers, and
 * an engine may offer one without the other.
 */

import type { BinaryNode } from 'zapo-js'
import { decodeBinaryNode, encodeBinaryNode } from 'zapo-js/transport'

/** Why a stanza could not be sent. */
export class SendError extends Error {
    constructor(
        message: string,
        /** The engine's own error, when there was one. */
        public override readonly cause?: unknown
    ) {
        super(message)
        this.name = 'SendError'
    }
}

/**
 * Raised when there is no live connection to send on.
 *
 * Separate from a plain {@link SendError} because it is the one a consumer can
 * act on without knowing anything about the engine: wait, reconnect, try again.
 */
export class NotConnectedError extends SendError {
    constructor(cause?: unknown) {
        super('not connected', cause)
        this.name = 'NotConnectedError'
    }
}

/** What this sender needs of the engine — the one call, nothing more. */
export interface NodeSender {
    readonly sendNode: (node: BinaryNode) => Promise<void>
}

/** What a requester needs: send, and be handed the reply. */
export interface NodeRequester extends NodeSender {
    readonly query: (node: BinaryNode, timeoutMs?: number) => Promise<BinaryNode>
}

/** Puts a stanza on the wire. */
export interface StanzaSender {
    /**
     * Send one stanza, as the frame bytes a decoder would consume.
     *
     * Resolving means the engine accepted the frame for delivery — not that the
     * server acted on it. Nothing at L0 can promise the latter: the answer to a
     * stanza is another stanza, and it arrives inbound.
     */
    readonly sendFrame: (frame: Uint8Array) => Promise<void>
}

/** Raised when a request produced no usable reply. */
export class RequestError extends Error {
    constructor(
        message: string,
        public override readonly cause?: unknown
    ) {
        super(message)
        this.name = 'RequestError'
    }
}

/**
 * Raised when the stanza left and nothing came back in time.
 *
 * Kept apart from a failed send because the two call for opposite responses: a
 * send that failed can be retried, while a request that timed out may well have
 * been acted on — retrying repeats whatever it did.
 */
export class RequestTimeoutError extends RequestError {
    constructor(cause?: unknown) {
        super('no reply before the deadline', cause)
        this.name = 'RequestTimeoutError'
    }
}

/** Sends a stanza and hands back the reply the server correlated to it. */
export interface StanzaRequester extends StanzaSender {
    /**
     * Send `frame` and wait for its reply, which crosses as a frame like
     * everything else — unparsed, because interpreting it is L1's job and a
     * consumer may want the bytes exactly as they arrived.
     */
    readonly requestFrame: (frame: Uint8Array, timeoutMs?: number) => Promise<Uint8Array>
}

/**
 * A requester over an engine's own `query`.
 *
 * A strictly stronger claim than {@link createSender}, and a separate capability
 * for that reason: correlating a reply means holding the engine's table of
 * outstanding requests, which an engine may not expose even when it will
 * happily write to the socket.
 */
export function createRequester(engine: NodeRequester): StanzaRequester {
    const sender = createSender(engine)
    return {
        sendFrame: sender.sendFrame,
        async requestFrame(frame: Uint8Array, timeoutMs?: number): Promise<Uint8Array> {
            let node: BinaryNode
            try {
                node = decodeBinaryNode(Buffer.from(frame))
            } catch (error) {
                throw new RequestError('frame is not a decodable stanza', error)
            }

            try {
                const reply = await engine.query(node, timeoutMs)
                return new Uint8Array(encodeBinaryNode(reply))
            } catch (error) {
                if (isTimeout(error)) {
                    throw new RequestTimeoutError(error)
                }
                throw isNotConnected(error)
                    ? new NotConnectedError(error)
                    : new RequestError('the request failed', error)
            }
        },
    }
}

/**
 * Whether the engine failed because nothing came back in time.
 *
 * Matched on the message, like {@link isNotConnected} and for the same reason:
 * `zapo` does not type this failure, and a miss degrades to a plain
 * {@link RequestError}, which still tells the caller the request failed.
 */
function isTimeout(error: unknown): boolean {
    const message = error instanceof Error ? error.message : String(error)
    return /timed? ?out|deadline/i.test(message)
}

/**
 * A sender over an engine's own `sendNode`.
 *
 * Takes the narrowest thing that works rather than the whole plugin context: a
 * seam this small is one a test can stand in for, and one a future `zapo`
 * reshuffle is unlikely to move.
 */
export function createSender(engine: NodeSender): StanzaSender {
    return {
        async sendFrame(frame: Uint8Array): Promise<void> {
            let node: BinaryNode
            try {
                node = decodeBinaryNode(Buffer.from(frame))
            } catch (error) {
                // Distinguished from an engine refusal on purpose: a frame this
                // adapter cannot read never reached the socket, and the fix is
                // in what the caller passed rather than in the connection.
                throw new SendError('frame is not a decodable stanza', error)
            }

            try {
                await engine.sendNode(node)
            } catch (error) {
                throw isNotConnected(error)
                    ? new NotConnectedError(error)
                    : new SendError('engine refused the send', error)
            }
        },
    }
}

/**
 * Whether the engine failed because nothing is connected.
 *
 * Matched on the message because `zapo` does not type this failure, and the
 * distinction is worth keeping: it is the only send failure a consumer can
 * retry its way out of. A miss degrades to a plain {@link SendError}, which is
 * the safe direction — the caller still learns the send failed.
 */
function isNotConnected(error: unknown): boolean {
    const message = error instanceof Error ? error.message : String(error)
    return /not connected|no connection|socket (is )?closed/i.test(message)
}

File: crates/wa-wire-adapter/src/send.rs
//! Sending a stanza, whichever engine is underneath.
//!
//! The inbound half of the boundary is a sink an adapter pushes into. This is
//! the other direction, and it is not symmetric: sending touches a socket, so
//! it is fallible and it takes time. That makes it the one part of the boundary
//! that has to be async, and the one that needs an error type consumers can
//! act on.
//!
//! # What crosses
//!
//! The same frame bytes as inbound — the buffer a decoder consumes. Deliberately
//! the same definition in both directions, because the useful consequence is
//! that a captured envelope can be sent back as it stands. Record and replay
//! stop being separate features.
//!
//! What an engine wants underneath differs (one takes marshalled bytes, another
//! a decoded node), and converting is the adapter's job. A consumer that had to
//! know which would be a consumer coupled to an engine.
//!
//! # What this is not
//!
//! Not request/response. Sending a stanza and correlating the reply the server
//! sends back is [`Capability::L0Request`], a separate claim — an engine can
//! perfectly well let you write to the socket without giving you its
//! correlation table.
//!
//! [`Capability::L0Request`]: wa_wire_contract::Capability::L0Request

extern crate alloc;

use alloc::boxed::Box;
use core::fmt;
use core::future::Future;
use core::pin::Pin;

/// A future an adapter's send returns.
///
/// Boxed because [`StanzaSender`] is used behind a trait object: a consumer
/// holds "some engine" and the whole point is not knowing which.
pub type SendFuture<'a> = Pin<Box<dyn Future<Output = Result<(), SendError>> + Send + 'a>>;

/// Why a stanza could not be sent.
///
/// Deliberately few variants. A consumer's realistic choices are retry, give up,
/// or reconnect, and a taxonomy finer than that would be an engine's error
/// vocabulary leaking through a boundary built to hide it.
#[derive(Debug)]
pub enum SendError {
    /// There is no live connection to send on.
    ///
    /// Distinct from [`Engine`](Self::Engine) because it is the one a consumer
    /// can act on without knowing anything about the engine: wait, reconnect,
    /// try again.
    NotConnected,
    /// The engine refused or failed the send.
    ///
    /// Carries the engine's own error so a report can name what happened, while
    /// keeping the type opaque so a consumer cannot come to depend on which
    /// engine produced it.
    Engine(Box<dyn core::error::Error + Send + Sync>),
}

impl fmt::Display for SendError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::NotConnected => f.write_str("not connected"),
            Self::Engine(error) => write!(f, "engine refused the send: {error}"),
        }
    }
}

impl core::error::Error for SendError {
    fn source(&self) -> Option<&(dyn core::error::Error + 'static)> {
        match self {
            Self::NotConnected => None,
            Self::Engine(error) => Some(error.as_ref()),
        }
    }
}

/// Puts a stanza on the wire.
///
/// The outbound half of what an adapter provides. An adapter that implements it
/// declares [`Capability::L0Outbound`], and a consumer that needs to send
/// requires that capability at setup rather than discovering at runtime that
/// its engine only listens.
///
/// [`Capability::L0Outbound`]: wa_wire_contract::Capability::L0Outbound
pub trait StanzaSender: Send + Sync {
    /// Send one stanza, as the frame bytes a decoder would consume.
    ///
    /// Resolving means the engine accepted the frame for delivery — not that
    /// the server acted on it. Nothing at L0 can promise the latter: the answer
    /// to a stanza is another stanza, and it arrives inbound.
    ///
    /// # Errors
    ///
    /// [`SendError::NotConnected`] when there is no live connection, and
    /// [`SendError::Engine`] for anything the engine itself refused.
    fn send_frame<'a>(&'a self, frame: &'a [u8]) -> SendFuture<'a>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::string::ToString;

    #[derive(Debug)]
    struct Refused;

    impl fmt::Display for Refused {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            f.write_str("socket closed")
        }
    }

    impl core::error::Error for Refused {}

    #[test]
    fn not_connected_reads_as_itself() {
        assert_eq!(SendError::NotConnected.to_string(), "not connected");
        assert!(core::error::Error::source(&SendError::NotConnected).is_none());
    }

    #[test]
    fn an_engine_failure_names_what_happened_and_keeps_the_cause() {
        let error = SendError::Engine(Box::new(Refused));

        assert_eq!(error.to_string(), "engine refused the send: socket closed");
        assert!(
            core::error::Error::source(&error).is_some(),
            "the cause is reachable for a report"
        );
    }

    /// A sender that records what it was handed, which is all a test of the
    /// boundary can check: what reaches the socket is the engine's business.
    ///
    /// Counts rather than collects, because [`StanzaSender`] is `Sync` and this
    /// crate is `no_std` — there is no `Mutex` to reach for, and the bytes are
    /// checked on the way in instead.
    #[derive(Default)]
    struct Recording {
        sent: core::sync::atomic::AtomicUsize,
        bytes: core::sync::atomic::AtomicUsize,
    }

    impl StanzaSender for Recording {
        fn send_frame<'a>(&'a self, frame: &'a [u8]) -> SendFuture<'a> {
            Box::pin(async move {
                self.sent
                    .fetch_add(1, core::sync::atomic::Ordering::Relaxed);
                self.bytes
                    .fetch_add(frame.len(), core::sync::atomic::Ordering::Relaxed);
                Ok(())
            })
        }
    }

    #[test]
    fn a_sender_is_usable_behind_a_trait_object() {
        // The property the boxed future exists for: a consumer holds "some
        // engine" and does not know which.
        let recording = Recording::default();
        let sender: &dyn StanzaSender = &recording;

        let sent = block_on(sender.send_frame(b"frame"));

        assert!(sent.is_ok());
        assert_eq!(
            recording.sent.load(core::sync::atomic::Ordering::Relaxed),
            1
        );
        assert_eq!(
            recording.bytes.load(core::sync::atomic::Ordering::Relaxed),
            5,
            "the frame crossed as it stands"
        );
    }

    /// Minimal executor: this crate has no runtime dependency and should not
    /// grow one for a test. The futures here never yield, so a no-op waker is
    /// all a poll loop needs.
    fn block_on<T>(future: impl Future<Output = T>) -> T {
        use core::task::{Context, Poll, Waker};

        let waker = Waker::noop();
        let mut context = Context::from_waker(waker);
        let mut future = Box::pin(future);
        loop {
            if let Poll::Ready(value) = future.as_mut().poll(&mut context) {
                return value;
            }
        }
    }
}
